File: src/Core.Go/tri_boolean/tri_boolean.go
package triboolean

import (
	"errors"
	"fmt"
	"math"
)

type State string

const (
	T State = "T"
	F State = "F"
	N State = "N"
)

type Tri struct {
	S State `json:"s"`
}

func FromBool(b bool) Tri {
	if b {
		return Tri{S: T}
	}
	return Tri{S: F}
}

func Held() Tri {
	return Tri{S: N}
}

func (t Tri) IsLiving() bool {
	return t.S == N
}

func (t Tri) IsCertain() bool {
	return t.S != N
}

func Cooperate(t Tri) Tri {
	return t
}

type MeasureFeedback struct {
	Reason string `json:"reason"`
}

type MeasureResult struct {
	Ok       bool             `json:"ok"`
	Value    bool             `json:"value,omitempty"`
	Feedback *MeasureFeedback `json:"feedback,omitempty"`
}

func Measure(t Tri) (bool, error) {
	switch t.S {
	case T:
		return true, nil
	case F:
		return false, nil
	default:
		return false, errors.New("collapsed-living-uncertainty")
	}
}

func MapTri(t Tri, fn func(bool) bool) Tri {
	switch t.S {
	case T:
		return FromBool(fn(true))
	case F:
		return FromBool(fn(false))
	default:
		return Tri{S: N}
	}
}

func BindTri(t Tri, fn func(bool) Tri) Tri {
	switch t.S {
	case T:
		return fn(true)
	case F:
		return fn(false)
	default:
		return Tri{S: N}
	}
}

func NotTri(t Tri) Tri {
	switch t.S {
	case T:
		return Tri{S: F}
	case F:
		return Tri{S: T}
	default:
		return Tri{S: N}
	}
}

func AndTri(a, b Tri) Tri {
	if a.S == F || b.S == F {
		return Tri{S: F}
	}
	if a.S == N || b.S == N {
		return Tri{S: N}
	}
	return Tri{S: T}
}

func OrTri(a, b Tri) Tri {
	if a.S == T || b.S == T {
		return Tri{S: T}
	}
	if a.S == N || b.S == N {
		return Tri{S: N}
	}
	return Tri{S: F}
}

type FloatShape struct {
	HighWidth    int `json:"highWidth"`
	DecoderWidth int `json:"decoderWidth"`
	LowWidth     int `json:"lowWidth"`
}

var DefaultShape = FloatShape{HighWidth: 4, DecoderWidth: 3, LowWidth: 4}

type TriFloat struct {
	Shape   FloatShape `json:"shape"`
	High    []Tri      `json:"high"`
	Decoder []Tri      `json:"decoder"`
	Low     []Tri      `json:"low"`
}

func IntOf(trits []Tri) (int, bool) {
	v := 0
	for _, t := range trits {
		if t.S == N {
			return 0, false
		}
		v = v * 2
		if t.S == T {
			v += 1
		}
	}
	return v, true
}

func IntToTrits(v int, width int) []Tri {
	out := make([]Tri, width)
	for i := width - 1; i >= 0; i-- {
		var state = F
		if ((v >> i) & 1) == 1 {
			state = T
		}
		out[width-1-i] = Tri{S: state}
	}
	return out
}

type FloatFeedback struct {
	Reason string `json:"reason"`
}

type DecodeResult struct {
	Ok       bool           `json:"ok"`
	Value    float64        `json:"value"`
	Feedback *FloatFeedback `json:"feedback,omitempty"`
}

func Decode(f TriFloat) DecodeResult {
	mode, ok := IntOf(f.Decoder)
	if !ok {
		return DecodeResult{Ok: false, Feedback: &FloatFeedback{Reason: "interpretation-superposed"}}
	}
	allTrits := append([]Tri{}, f.High...)
	allTrits = append(allTrits, f.Low...)
	v, ok := IntOf(allTrits)
	if !ok {
		return DecodeResult{Ok: false, Feedback: &FloatFeedback{Reason: "value-superposed"}}
	}
	bias := 1 << (f.Shape.DecoderWidth - 1)
	exponent := mode - bias
	return DecodeResult{Ok: true, Value: float64(v) * math.Pow(2, float64(exponent))}
}

type EncodeFeedback struct {
	Reason string `json:"reason"`
	Detail string `json:"detail"`
}

type EncodeResult struct {
	Ok       bool            `json:"ok"`
	Float    *TriFloat       `json:"float,omitempty"`
	Feedback *EncodeFeedback `json:"feedback,omitempty"`
}

func FromValue(value float64, shape FloatShape) EncodeResult {
	if math.IsNaN(value) || math.IsInf(value, 0) || value < 0 {
		return EncodeResult{Ok: false, Feedback: &EncodeFeedback{Reason: "not-representable", Detail: "v0 is unsigned + finite"}}
	}
	valueBits := shape.HighWidth + shape.LowWidth
	maxMode := (1 << shape.DecoderWidth) - 1
	maxV := 1 << valueBits
	bias := 1 << (shape.DecoderWidth - 1)

	for mode := 0; mode <= maxMode; mode++ {
		scaled := value / math.Pow(2, float64(mode-bias))
		if scaled == math.Floor(scaled) && scaled >= 0 && scaled < float64(maxV) {
			bits := IntToTrits(int(scaled), valueBits)
			return EncodeResult{
				Ok: true,
				Float: &TriFloat{
					Shape:   shape,
					High:    bits[:shape.HighWidth],
					Decoder: IntToTrits(mode, shape.DecoderWidth),
					Low:     bits[shape.HighWidth:],
				},
			}
		}
	}
	return EncodeResult{
		Ok: false,
		Feedback: &EncodeFeedback{
			Reason: "not-representable",
			Detail: fmt.Sprintf("no (mode,V) with mode<=%d and V<%d represents %v", maxMode, maxV, value),
		},
	}
}

func FromTrits(high []Tri, decoder []Tri, low []Tri) TriFloat {
	return TriFloat{
		Shape:   FloatShape{HighWidth: len(high), DecoderWidth: len(decoder), LowWidth: len(low)},
		High:    high,
		Decoder: decoder,
		Low:     low,
	}
}

File: src/Core.Python/src/zeta/tri_boolean.py
from collections.abc import Callable
from typing import Literal

State = Literal["T", "F", "N"]


class Tri:
    __slots__ = ("s",)

    def __init__(self, s: State):
        self.s = s

    def __eq__(self, other):
        if not isinstance(other, Tri):
            return False
        return self.s == other.s

    def __hash__(self):
        return hash(self.s)

    def __repr__(self):
        return f"Tri({self.s!r})"

    def is_living(self) -> bool:
        return self.s == "N"

    def is_certain(self) -> bool:
        return self.s != "N"


T = Tri("T")
F = Tri("F")
N = Tri("N")


def from_bool(b: bool) -> Tri:
    return T if b else F


def held() -> Tri:
    return N


def cooperate(t: Tri) -> Tri:
    return t


class MeasureFeedback:
    def __init__(self, reason: str):
        self.reason = reason


class MeasureResult:
    def __init__(
        self,
        ok: bool,
        value: bool = False,
        feedback: MeasureFeedback | None = None,
    ):
        self.ok = ok
        self.value = value
        self.feedback = feedback


def measure(t: Tri) -> MeasureResult:
    if t.s == "T":
        return MeasureResult(ok=True, value=True)
    elif t.s == "F":
        return MeasureResult(ok=True, value=False)
    else:
        return MeasureResult(
            ok=False, feedback=MeasureFeedback(reason="collapsed-living-uncertainty")
        )


def map_tri(t: Tri, fn: Callable[[bool], bool]) -> Tri:
    if t.s == "T":
        return from_bool(fn(True))
    elif t.s == "F":
        return from_bool(fn(False))
    return N


def bind_tri(t: Tri, fn: Callable[[bool], Tri]) -> Tri:
    if t.s == "T":
        return fn(True)
    elif t.s == "F":
        return fn(False)
    return N


def not_tri(t: Tri) -> Tri:
    if t.s == "T":
        return F
    elif t.s == "F":
        return T
    return N


def and_tri(a: Tri, b: Tri) -> Tri:
    if a.s == "F" or b.s == "F":
        return F
    if a.s == "N" or b.s == "N":
        return N
    return T


def or_tri(a: Tri, b: Tri) -> Tri:
    if a.s == "T" or b.s == "T":
        return T
    if a.s == "N" or b.s == "N":
        return N
    return F


class FloatShape:
    def __init__(self, high_width: int, decoder_width: int, low_width: int):
        self.high_width = high_width
        self.decoder_width = decoder_width
        self.low_width = low_width


DEFAULT_SHAPE = FloatShape(high_width=4, decoder_width=3, low_width=4)


class TriFloat:
    def __init__(
        self,
        shape: FloatShape,
        high: tuple[Tri, ...],
        decoder: tuple[Tri, ...],
        low: tuple[Tri, ...],
    ):
        self.shape = shape
        self.high = high
        self.decoder = decoder
        self.low = low


def int_of(trits: tuple[Tri, ...]) -> int | None:
    v = 0
    for t in trits:
        if t.s == "N":
            return None
        v = v * 2 + (1 if t.s == "T" else 0)
    return v


def int_to_trits(v: int, width: int) -> tuple[Tri, ...]:
    out = []
    for i in range(width - 1, -1, -1):
        out.append(T if ((v >> i) & 1) == 1 else F)
    return tuple(out)


class FloatFeedback:
    def __init__(self, reason: str):
        self.reason = reason


class DecodeResult:
    def __init__(
        self, ok: bool, value: float = 0.0, feedback: FloatFeedback | None = None
    ):
        self.ok = ok
        self.value = value
        self.feedback = feedback


def decode(f: TriFloat) -> DecodeResult:
    mode = int_of(f.decoder)
    if mode is None:
        return DecodeResult(
            ok=False, feedback=FloatFeedback(reason="interpretation-superposed")
        )
    v = int_of(f.high + f.low)
    if v is None:
        return DecodeResult(ok=False, feedback=FloatFeedback(reason="value-superposed"))
    bias = 2 ** (f.shape.decoder_width - 1)
    exponent = mode - bias
    return DecodeResult(ok=True, value=float(v * (2**exponent)))


def cooperate_float(f: TriFloat) -> TriFloat:
    return f


def is_held(f: TriFloat) -> bool:
    return not decode(f).ok


class EncodeFeedback:
    def __init__(self, reason: str, detail: str):
        self.reason = reason
        self.detail = detail


class EncodeResult:
    def __init__(
        self,
        ok: bool,
        float_val: TriFloat | None = None,
        feedback: EncodeFeedback | None = None,
    ):
        self.ok = ok
        self.float = float_val
        self.feedback = feedback


def from_value(value: float, shape: FloatShape = DEFAULT_SHAPE) -> EncodeResult:
    import math

    if not math.isfinite(value) or value < 0:
        return EncodeResult(
            ok=False,
            feedback=EncodeFeedback(
                reason="not-representable", detail="v0 is unsigned + finite"
            ),
        )
    value_bits = shape.high_width + shape.low_width
    max_mode = (1 << shape.decoder_width) - 1
    max_v = 1 << value_bits
    bias = 2 ** (shape.decoder_width - 1)

    for mode in range(max_mode + 1):
        scaled = value / (2 ** (mode - bias))
        if scaled.is_integer() and scaled >= 0 and scaled < max_v:
            bits = int_to_trits(int(scaled), value_bits)
            return EncodeResult(
                ok=True,
                float_val=TriFloat(
                    shape=shape,
                    high=bits[: shape.high_width],
                    decoder=int_to_trits(mode, shape.decoder_width),
                    low=bits[shape.high_width :],
                ),
            )
    return EncodeResult(
        ok=False,
        feedback=EncodeFeedback(
            reason="not-representable",
            detail=f"no (mode,V) with mode<={max_mode} and V<{max_v} represents {value}",
        ),
    )


def from_trits(
    high: tuple[Tri, ...], decoder: tuple[Tri, ...], low: tuple[Tri, ...]
) -> TriFloat:
    shape = FloatShape(
        high_width=len(high), decoder_width=len(decoder), low_width=len(low)
    )
    return TriFloat(shape=shape, high=high, decoder=decoder, low=low)
